### PR TITLE
Feature/pytest - translate testing to pytest

### PR DESCRIPTION
**Pull request by samuelhward**
- convert from unittest to pytest (simpler, more capable)
- added template file for custom exceptions
- updated docs to reflect
- addresses #75 #70

### CODE REVIEW

File: pyfecons/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import pyfecons
+
+class TemplateException(Exception):
```
Comment (craastad) [inline pyfecons/exceptions.py:3]: What's the purpose of this class? Do you intend to add more tests here?

File: requirements.txt [hunk 2]
```diff
@@ -1,6 +1,10 @@
+# top level requirements can be generated with
+# pipdeptree -fd 0
+
```
Comment (craastad) [inline requirements.txt:3]: Why are these lines here? I think the requirements file should be as clean as possible.

File: tests/test_tests.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
+# placeholder file to test the testing capability
```
Comment (craastad) [inline tests/test_tests.py]: Comment is incorrect. This is an actual test file for these functions that were exposed for nttau AI optimization (though they ended up writing their own function).

File: tests/test_tests.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    BlanketPrimaryCoolant,
+    BlanketSecondaryCoolant,
+    BlanketNeutronMultiplier,
+    BlanketStructure,
+)
+from pyfecons.inputs import Blanket, RadialBuild
+from pyfecons.units import Ratio, Meters
+
+
+def test_mfe_tokamak():
```
Comment (craastad) [inline tests/test_tests.py]: Since we removed the TestCAS220101TotalCost class, I think it would be nice for the naming to reflect that this is just testing the cost calculation for CAS 22.1.1. Maybe `test_mfe_tokamak_cas220101_total_cost`?

File: tests/test_tests.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    result_cost = compute_cas22_reactor_equipment_total_cost(
+        ReactorType.MFE, ConfinementType.MAGNETIC_MIRROR, blanket, radial_build
+    )
+    expected_result_cost = (
+        1217.12  #! this could become deprecated if cost calculation changes
+    )
+    assert pytest.approx(result_cost, rel=1e-2) == expected_result_cost
+
+
+def test_mfe_mirror_minimal():
```
Comment (craastad) [inline tests/test_tests.py]: Same more, descriptive test name.

File: tests/test_tests.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
+# placeholder file to test the testing capability
+
+import pytest
+
```
Comment (craastad) [inline tests/test_tests.py]: Can we have a more descriptive name for the test file? The old name better described the purpose of the file, testing the cas 22.1.1 cost functions `test_cas220101_cost_functions.py`

File: pyfecons/tests/test_cas220101_cost_functions.py [hunk 7]
```diff
@@ -1,100 +0,0 @@
-import unittest
```
Comment (craastad) [inline pyfecons/tests/test_cas220101_cost_functions.py:1]: In the future make sure to have files moved rather than delete + add to preserve git history. Especially in active code.

**Review by craastad — APPROVED** [review]
I'm glad pytest is more concise than unittest so thanks for adding it :) 

Please consider comments, especially around the test file name becoming less descriptive and removing extra comments / code.

File: pyfecons/exceptions.py (hunk 1 above)
Comment (samuelhward) [reply pyfecons/exceptions.py:3]: Yes, it's just a placeholder to mark where new exceptions should be defined; once we have implemented a new exception this line can be removed.

File: requirements.txt (hunk 2 above)
Comment (samuelhward) [reply requirements.txt:3]: I noticed requirements.txt only featured top-level dependencies, so I included a comment showing users how to generate a requirements.txt file in this style. Can be removed.

File: tests/test_tests.py (hunk 3 above)
Comment (samuelhward) [reply tests/test_tests.py]: I see - I read "These are more of demonstration tests of what's implemented rather than testing the code works at a fine detail." in the top-level README.md and so thought the tests were just placeholders/demonstrations. 

Feel free to suggest a new comment.

File: tests/test_tests.py (hunk 4 above)
Comment (samuelhward) [reply tests/test_tests.py]: Done

File: tests/test_tests.py (hunk 5 above)
Comment (samuelhward) [reply tests/test_tests.py]: Done

File: tests/test_tests.py (hunk 6 above)
Comment (samuelhward) [reply tests/test_tests.py]: Done

File: pyfecons/tests/test_cas220101_cost_functions.py (hunk 7 above)
Comment (samuelhward) [reply pyfecons/tests/test_cas220101_cost_functions.py:1]: oh nice! didn't know about this

File: requirements.txt (hunk 2 above)
Comment (craastad) [reply requirements.txt:3]: Got it. My intention with requirements.txt is having the canonical dependencies. I know listing the dependencies and adding them to the requirement.txt file usually has so many dependencies it's hard to tell what are the "actual" dependencies.

Doesn't hurt to keep it in.

**Review by craastad — APPROVED** [review]
Thanks for addressing comments. :)

**Comment by samuelhward** [comment]
@craastad gotcha on dependencies - how do you generate the top level out of curiosity? If we want a clean file (since it will be overwritten with pip freeze anyway), then it might be worth adding a style guide where we define how to contribute to these files.

We could also add automatic formatting e.g. Black too, to ensure consistency over files.
